### PR TITLE
Implement generic number literals

### PR DESCRIPTION
**Pull request by odersky**
Example:
```
val x: BigInt = 111111100000022222222222
```
 - Also works for user defined types.
 - Conversion can be done at compile time, with customizable errors.

### CODE REVIEW

**Comment by propensive** [comment]
I'm very happy to see this! Are there plans to provide support for pattern matching too?

**Comment by odersky** [comment]
> I'm very happy to see this! Are there plans to provide support for pattern matching too?

I have not thought about it. Do you have a proposal how to do that?

**Comment by odersky** [comment]
Regarding pattern matching, if the expansion of a literal creates something matchable (such a case class instance) we are good. The problem is what to do if that's not the case. Right now, we treat an application `f(e)` in a pattern always as a constructor pattern that is decomposed with `unapply`. We would need a syntax where it gets treated as a plain function call instead. I.e. something like:
```
  case {f(e)} => ...
```
If we have that, the inline expansion of a number literal can add the braces (or whatever) to make it work. So the problem is really more general than just literals.

**Comment by propensive** [comment]
I had only ever thought about it casually, but the `case {f(e)} => ...` syntax looks interesting. I'm yet to discover whether Scala 3 has sufficient macro support for me to reimplement [Kaleidoscope](https://github.com/propensive/kaleidoscope), to allow pattern matching on regular expressions, e.g.
```
str match {
  case r"$firstName([a-zA-Z])* $lastName@([a-zA-Z]*)" => (firstName, lastName)
  case _ => ("Bad", "Name")
}
```
(which was probably the most obscure feature I ever used in Scala 2.)

but it's quite high up my list of desirable features. Relatedly, I also often thought it would be nice to be able to explicitly specify types to extractors, like so:
```
str match {
  case As[Int](int) => int
  case _ => 0
}
```

I'll think about the general idea some more.

**Comment by odersky** [comment]
It turns out that the idea to treat contents of blocks in patterns as expressions works with some minor tweaks. So this means we support generic literals as patterns now. So far, this is all internal; we do not support `{...}` as surface syntax yet. Whether we should do this is another discussion to have.

**Comment by propensive** [comment]
That sounds like it can open up a whole variety of interesting possibilities. Thank you!

File: docs/docs/reference/changed-features/numeric-literals.md [hunk 1]
```diff
@@ -0,0 +1,245 @@
...
+1e10_0000_000_000: BigFloat
+```
+would be expanded by the compiler to
+```scala
+BigFloat.FromDigits.fromDigits("1e10_0000_000_000")
+```
+Evaluating this expression throws a `NumberTooLarge` exception at run time. We would like it to
+produce a compile-time error instead. We can achieve this by tweaking the `BigFloat` class
+with a small dose of meta-programming. The idea is to turn the `fromDigits` method
+of the into a macro, i.e. make it an inline method with a splice as right hand side.
```
Comment (propensive) [inline docs/docs/reference/changed-features/numeric-literals.md]: of the...?

**Comment by propensive** [comment]
@odersky Do you think there is any use in tweaking the parser to support multiple adjacent `.`s in number literals, in order to provide nicer syntax for integer ranges?
```
for(i <- 1..100) println("")
```

This would obviate `to`, and the `..` would have higher precedence than the alphanumeric infix method, which would be more convenient in a few cases.

**Comment by odersky** [comment]
> @odersky Do you think there is any use in tweaking the parser to support multiple adjacent .s in number literals, in order to provide nicer syntax for integer ranges?

I like `to` and `until`. `..` by itself is not enough since it does not let you do half open intervals.

**Comment by odersky** [comment]
@anatoliykmetyuk @nicolasstucki Can you figure out why the CI does not pass? Everything works fine locally.

**Comment by anatoliykmetyuk** [comment]
@odersky `toExpr` has changed its signature, the correct implementation is:

```scala
  given as Liftable[BigInt] {
    def toExpr(x: BigInt): given (qctx: QuoteContext) => Expr[BigInt] =
      '{BigInt(${x.toString.toExpr})}
  }
```

It should also fail locally. However, I encountered troubles recompiling the PR: even after doing `all:clean`, I had some broken classes with broken tasty signature. `rm -rf out/bootstrap/` helped me; I think if you do `rm -rf out/` there's a chance it will also start failing locally for you.

**Comment by nicolasstucki** [comment]
Or slightly more compact 
```scala
  given as Liftable[BigInt] {
    def toExpr(x: BigInt) = '{ BigInt(${x.toString.toExpr}) }
  }
```

**Comment by nicolasstucki** [comment]
May also need a rebase to have the 0.17.0-RC1 reference compiler.

File: compiler/src/dotty/tools/dotc/typer/Typer.scala [hunk 2]
```diff
@@ -703,7 +767,8 @@ class Typer extends Namer
     (index(stats), typedStats(stats, ctx.owner))
 
   def typedBlock(tree: untpd.Block, pt: Type)(implicit ctx: Context): Tree = track("typedBlock") {
-    val (exprCtx, stats1) = typedBlockStats(tree.stats)
+    val localCtx = ctx.retractMode(Mode.Pattern)
```
Comment (nicolasstucki) [inline compiler/src/dotty/tools/dotc/typer/Typer.scala]: Where did this `Mode.Pattern` come from? 
Was it from the pattern of something like?
```
bigFloat match {
  case 123_344_537_244_453E433 => ???
}
```

File: compiler/src/dotty/tools/dotc/typer/Typer.scala [hunk 3]
```diff
@@ -508,6 +509,69 @@ class Typer extends Namer
...
+        val fromDigitsCls = tree.kind match {
+          case Whole(10) => defn.FromDigitsClass
+          case Whole(_) => defn.FromDigits_WithRadixClass
+          case Decimal => defn.FromDigits_DecimalClass
+          case Floating => defn.FromDigits_FloatingClass
+        }
+        inferImplicit(fromDigitsCls.typeRef.appliedTo(target), EmptyTree, tree.span) match {
+          case SearchSuccess(arg, _, _) =>
+            val fromDigits = untpd.Select(untpd.TypedSplice(arg), nme.fromDigits).withSpan(tree.span)
+            val firstArg = Literal(Constant(digits.toString))
```
Comment (nicolasstucki) [inline compiler/src/dotty/tools/dotc/typer/Typer.scala]:
```suggestion
            val firstArg = Literal(Constant(digits))
```

File: compiler/src/dotty/tools/dotc/typer/Typer.scala (hunk 2 above)
Comment (odersky) [reply compiler/src/dotty/tools/dotc/typer/Typer.scala]: Yes, exacty. We might make it user-accessible syntax though.

**Comment by odersky** [comment]
The main open issue is what the right typeclass infrastructure should be. 

**Question 1:** Should `FromDigits` and `FromString` be merged?

Merging is attractive because it simplifies the design but in summary I believe the two classes are better kept separate. The contract for `FromDigits` is different since we only need to be ready to parse strings that are numbers. That means implementations of `FromDigits` can dispense with a number of checks that implementations of `FromString` have to perform. 

**Question 2**: Should we collapse the four classes `FromDigits`, `FromDigits.WithRadix`, `FromDigits.Decimal` and `FromDigits.Floating` into a single one? 

Again, we trade off conceptual simplicity for ease of implementation. Here the choice actually does make a difference. Say, we have class FooNum that implements `FromDigits` but not `FromDigits.Decimal`. We write
```scala
val x: FooNum = 2.0
```
Under the current 4-class scenario, `2.0` would be treated as a floating point number. The program might still typecheck if there is an implicit conversion from Doubles to FooNums. 

Under a collapsed scenario, 2.0 would be treated as a `FooNum` and the conversion would give an error.

So the current 4-class scheme is closest to the status quo; it only treats a literal as a user defined type if the user defined type has declared that it accepts that kind of literal.

**Comment by odersky** [comment]
In summary, I propose to stick with the status quo and merge it now. We can still change it later if different arguments come to light.

**Comment by sjrd** [comment]
I'm not convinced by this feature.

For starters, what is really the motivation?

The user-defined literals are expected-type-driven. This means that they work in some cases, but not in others. In particular:
* Somehow this works for pattern matching, but I don't see it working with `==`, i.e., I don't see how `someCustomNumberValue == 5` is going to do the right thing.
* It will probably work at the right-hand-side of operators, such as `someCustomNumber + 5`, because the expected type of `def +(arg)` will be `CustomNumber`, but it won't work at the left-hand-side of operators.

Can I have actual *constant expressions* of type `Byte` and `Short` with this feature? Currently it is impossible because `5.toByte` is not a constant expression, and `(5: Byte)` either, and that impossibility is a known problem in Scala.

Why does this work for numeric literals, but not for other literals? In particular, could I recover symbol "literals" from string literals with something like this?

Why is this considered when we already have string interpolators and we could write `bi"10_000_000_000"` instead?

Why is this feature considered morally better than implicit conversions, which are basically deprecated?

**Comment by odersky** [comment]
> i.e., I don't see how someCustomNumberValue == 5 is going to do the right thing.

Just define an overloaded equality `==` in the type of `someCustomerValue`.

> It will probably work at the right-hand-side of operators, such as someCustomNumber + 5, because the expected type of def +(arg) will be CustomNumber, but it won't work at the left-hand-side of operators.

Yes, that looks like a hard limitation. 

> Can I have actual constant expressions of type Byte and Short with this feature? 

No. Primitive number types are treated as before. But `5.toByte` can already now be a constant expression if `toByte` is an inline method.
 
> bi"10_000_000_000"

Simplicity. I find that ring interpolator really ugly. It's one of the clear advantages of languages with unlimited precision numbers that you can define numbers without artificial restrictions. Let's face it: the whole way we define numeric literals with suffixes and so on is artificial and unnatural. A literal is just what it is. It is given meaning by ascribing a type.

**Comment by sjrd** [comment]
> Just define an overloaded equality == in the type of someCustomerValue.

That has the same problem with the left-hand-side: `5 == someCustomNumberValue` will not be equivalent to `someCustomNumberValue == 5`.

> Let's face it: the whole way we define numeric literals with suffixes and so on is artificial and unnatural.

I disagree. The whole way we have different types of numbers with differing precisions is artificial and unnatural. It goes against math. But given that we have different types of numbers with different behaviors (based on ad hoc polymorphism to boot!), I'm very happy we have to use explicitly typed literals. They prevent potential warts.

**Comment by odersky** [comment]
I believe the examples with `5` as the number are all misleading. We already know how to deal with small numbers and custom types using implicit conversion. I.e.
```
5 + myBigInt
```
and 
```
myBigInt + 5
```
both work and can be made to work by library code for arbitrary number types. 

So the main improvement that this PR brings is that we can write large number literals which are not constrained by the ranges of existing number types. However, in practice you'd rarely write a large number literal inline. Common practice recommends that you define a constant instead. So, with this proposal it's
```scala
val x: Int = 100
val y: Long = 10_000_000_000
val z: BigInt = 10_000_000_000_000_000_000_000
```
instead of the status quo:
```
val x: Int = 100
val y: Long = 10_000_000_000L
val z: BigInt = bi"10_000_000_000_000_000_000_000"
```
I find the first much more pleasant than the second. The limitations when you can or cannot use a really large number inline in an expression are very similar to mixing existing small number literals and custom types. There is as far as I can see only one restriction for large literals:
```
50_000_000_000_000_000_000_000 + myBigInt
```
will not work. It works for small literals on the left since there are implicit conversions from `Int` and `Long` to `{ + : BigInt => ? }`. The conversion trick does not extend to typing literals. But overall I think that's a small restriction and we can live with it.

**Comment by odersky** [comment]
I should also note that it's not my idea. I got this from Guy Steele, who proposed this some time ago.

**Comment by smarter** [comment]
An extra concern in Scala is that so far, all literals have a corresponding singleton literal type (symbol literals are deprecated so I'll ignore them), but with this proposal, this would no longer be true. It's not necessarily a deal-breaker but I think it's worth exploring alternative designs that would support singleton types.

**Comment by denisrosset** [comment]
Amazing idea. Spire needs a lot of complex/hackish code to help people write mathematical like notation and it works 85% of the time (especially with all the Int implicit conversions already in the stdlib).

I'll take a closer look later on this.

**Comment by odersky** [comment]
I should not that both Java and Scala already do this to some degree: An `Int` literal is treated as a `Byte`, `Short`, or `Char` if the expected type is one of these and the literal fits the range. This is not an implicit conversion (there is none between `Int` and `Byte` or the other types), but a specialized treatment of literals. In a sense the proposal is to take that idea and apply it to user defined number types.

**Comment by AleksanderBG** [comment]
It seems to me that this feature simply adds more irregularity to the language than it gets rid of. The current status quo, as I understand it, is either defining a string interpolator and using it everywhere, or the complex/hackish implicit conversions that @denisrosset mentioned. Since I do not actually write any code that uses this feature, I will refrain from having any opinion, but I will point out what will not work in the current version of the proposal:

Equality comparisons with polymorphic literals will not work. If we're handling math vectors, then we may want to treat `1` and `0` as vectors. Then something like `val vec: Vec3 = 1` would work, but `vec == 1` would fail, no matter what we do - the `equals(Any)` method is always available, and will force the polymorphic literal into `Int` according to the proposed rules.

In general, binary operators will not work. If I defined a custom `MyBigInt` that can only be multiplied by other `MyBigInt`s, then doubling it like `2 * myBigInt` will fail.

Defining constants with expressions will not work:
```scala
val tau: BigFloat = 2 * 3.1415926535897932384626433832795028841971
```
That is, the "natural" way of defining `tau` fails to compile. Even worse, since the RHS will default to Double if it is small enough, expressions like that may on occasion compile and produce an imprecise result.

Two notable languages which do have polymorphic literals are Haskell and Rust. However, both of them also have complete local type inference, which avoids all of the above problems and makes the feature work that much better.
In particular, polymorphic literals are particularly pleasing in Rust, which differentiates between signed/unsigned and 32/64-bit integers, but where we do not need to care about that at all when using local literals:

```rust
fn print(i: i64) {
    println!("{}", i)
}

fn main() {
    let is = vec![1, 2, 3];
    for i in is {
        print(i);
    }
}
```

No matter whether `print` expects an `i64` or an `u32` or anything else, this code will compile (note that the syntax for monomorphic literals in Rust is `1u32`). See it [here](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=9aafb4d0722442fac8a6ad5ec768154e). Similar code works just as well in Haskell.

**Comment by rjolly** [comment]
Regarding equality comparison, we have to look how it plays with [multiversal equality](http://dotty.epfl.ch/docs/reference/contextual/multiversal-equality.html). Regarding non-commuting binary operators, something that would be really useful is if we could define a left operator with the colon notation:
`def +:(that: BigInt)`
, and use it without the colon, as in `that + this` instead of `that +: this`.

**Comment by propensive** [comment]
> Note that both Java and Scala already do this to some degree: An `Int` literal is treated as a `Byte`, `Short`, or `Char` if the expected type is one of these and the literal fits the range. This is not an implicit conversion (there is none between `Int` and `Byte` or the other types), but a specialized treatment of literals. In a sense the proposal is to take that idea and apply it to user defined number types.

As a suggestion for educational materials, instead of using the phrase "`Int` literal", could we call it an "integral literal" to indicate that it's just as much an `Int` literal as a `Byte` literal and a `Short` literal and that, in the absence of any other expected type, the compiler will default to typing it as an `Int`?

**Comment by denisrosset** [comment]
@odersky What is the merge timeline for this? I'd love to try this on a "minispire" library and provide feedback, but I'm quite overwhelmed by academic collaborations at the moment.

**Comment by odersky** [comment]
@denisrosset I'd like to merge this by the end of the month, at the latest. But even after merge it still has to be discussed in the SIP process.

**Pull request merged**